File: 2024/src/day10.py
from utils import get_digits

def find_ends(m, h, w, x, y, ends):
    elevation = m[y][x]
    if elevation == 9:
        ends[(x, y)] = True
        return 1

    count = 0
    if x > 0 and m[y][x - 1] == elevation + 1:
        count += find_ends(m, h, w, x - 1, y, ends)
    if x < w - 1 and m[y][x + 1] == elevation + 1:
        count += find_ends(m, h, w, x + 1, y, ends)
    if y > 0 and m[y - 1][x] == elevation + 1:
        count += find_ends(m, h, w, x, y - 1, ends)
    if y < h - 1 and m[y + 1][x] == elevation + 1:
        count += find_ends(m, h, w, x, y + 1, ends)
    return count

def day10(file):
    m = [get_digits(line) for line in file if line != "\n"]
    w = len(m[0])
    h = len(m)

    sol1 = 0
    sol2 = 0
    for y in range(0, h):
        for x in range(0, w):
            if m[y][x] != 0:
                continue
            ends = {}
            sol2 += find_ends(m, h, w, x, y, ends)
            sol1 += len(ends)

    return (sol1, sol2)

File: 2025/src/day10.rs
use aoc_lib_rust::{Day, Example, Solution};
use microlp::{Problem, OptimizationDirection, ComparisonOp};

pub struct Day10;

impl Day for Day10 {

    const PART1: Solution = Solution::U32(558);
    const PART2: Solution = Solution::U32(20317);

    fn solve(input: &str) -> [Solution; 2] {
        let mut sol1 = 0;
        let mut sol2 = 0;
        for line in input.lines() {
            let mut parts = line.split(' ');
            let diagram = parts.next().unwrap();
            let diagram = diagram.as_bytes().iter().skip(1).take(diagram.len() - 2)
                .rfold(0, |acc, x| (acc << 1) | !(*x == b'.') as u16);

            let mut buttons = Vec::new();
            let mut counters = Vec::new();
            for part in parts {
                if part.starts_with('{') {
                    counters = part[1..part.len() - 1].split(',')
                        .map(|n| n.parse::<u16>().unwrap())
                        .collect::<Vec<_>>();
                    break;
                }
                let button = part[1..part.len() - 1].split(',')
                    .map(|n| n.parse::<u16>().unwrap())
                    .fold(0, |acc, n| acc | (1 << n));
                buttons.push(button);
            }
            assert!(buttons.len() <= 16);
            assert!(counters.len() <= 16);

            let mut best = u32::MAX;
            for presses in 0..1_u16 << buttons.len() {
                let state = buttons.iter().enumerate().fold(0, |acc, (i, button)| {
                    acc ^ (button * ((presses >> i) & 1))
                });
                if state == diagram {
                    best = best.min(presses.count_ones());
                }
            }
            sol1 += best;

            let mut problem = Problem::new(OptimizationDirection::Minimize);
            let vars = (0..buttons.len())
                .map(|_| problem.add_integer_var(1.0, (0, i32::MAX)))
                .collect::<Vec<_>>();
            for (i, count) in counters.iter().enumerate() {
                problem.add_constraint(
                    buttons.iter().enumerate().map(|(j, b)| {
                        (vars[j], ((b & (1 << i)) != 0) as u8 as _)
                    }),
                    ComparisonOp::Eq,
                    *count as _,
                );
            }
            let solution = problem.solve().unwrap();
            sol2 += solution.objective().round() as u32;
        }

        [Solution::U32(sol1), Solution::U32(sol2)]
    }

    const EXAMPLES: &'static [Example] = &[
        Example {
            solution: [Solution::U32(7), Solution::U32(33)],
            input: "[.##.] (3) (1,3) (2) (2,3) (0,2) (0,1) {3,5,4,7}
[...#.] (0,2,3,4) (2,3) (0,4) (0,1,2) (1,2,3,4) {7,5,12,7,2}
[.###.#] (0,1,2,3,4) (0,3,4) (0,1,2,4,5) (1,2) {10,11,11,5,10,5}",
        },
    ];
}
